### PR TITLE
Prod db migration scripts

### PR DESCRIPTION
**Pull request by uolter**
### Type of changes

- [ ] Add new resources
- [x] Update configuration to existing resources
- [ ] Remove existing resources

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [x] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [x] No

### Other information

- [x] 1. Replicate (backup … db production) **cstar-p-postgresql-backup**: Owner PagoPa. ~38 mins.
- [x] 2. Detach the replica database from 1: it becomes a master copy.
- [x] 3. Delete the production databade **cstar-p-postgresql**: `you can't have two databases with the same name.` Owner PagoPa
- [x] 4. Replicate the source database. Call it **cstar-p-postgresql** for production (**cstar-u-posgresql** UAT): Owner SIA. This step can be done few day before the migration.
- [x] 5. Detach the replica database from 4: it becomes a master copy: Owner SIA
- [x] 6. Create a new subscription: Owner SIA.
- [x] 7. Move the database server **cstar-p-postgresql** from 5 to the new subscription created at 6: Owner SIA
- [x] 8. Move the subscriprion crated at 6. to the **pagopa.it** directory: Onwer SIA
- [x] 9. Move the database **cstar-p-postgresql** (from step 8) to the subscription **PROD-CSTAR** (88c709b0-11cf-4450-856e-f9bf54051c1d) resource group **cstar-p-db-rg**: Owner PagoPa 
- [x] 10. run the command 
 ```bash
 /terraform.sh import prod module.postgresql.azurerm_postgresql_server.this /subscriptions/88c709b0-11cf-4450-856e-f9bf54051c1d/resourceGroups/cstar-p-db-rg/providers/Microsoft.DBforPostgreSQL/servers/cstar-p-postgresql
 ```
- [x]  11. Run terraform apply to create all other resources with dependencies  
- [x]  12. ~~Change the admin password in **cstar-p-posgresql**: Owner **PagoPa**~~ Already managed by terraform
- [x]  13. ~~Truncate all flyway history tables. (porgres, bpd, rtd)~~ Not required: SIA does not have flyway tables
- [x]  14. Run the script `./flyway.sh migrate PROD-CSTAR postgres` to set database passwords: Owner PagoPa.
- [x]  15. Run flyway scripts to set dblinks 

---

### Production Database Migration scripts  

```
./flyway.sh clean PROD-CSTAR postgres
./flyway.sh baseline PROD-CSTAR postgres -baselineVersion=2
./flyway.sh migrate PROD-CSTAR postgres
./flyway.sh info PROD-CSTAR postgres
./flyway.sh clean PROD-CSTAR bpd
./flyway.sh baseline PROD-CSTAR bpd -baselineVersion=3
./flyway.sh migrate PROD-CSTAR bpd
./flyway.sh clean PROD-CSTAR rtd
./flyway.sh baseline PROD-CSTAR rtd -baselineVersion=2
./flyway.sh migrate PROD-CSTAR rtd
```
### If PR is partially applied, why? (reserved to maintainers)

We are waiting for **SIA** to complete the steps they own.

### CODE REVIEW

File: src/env/prod/terraform.tfvars [hunk 1]
```diff
@@ -217,7 +217,7 @@ cidr_subnet_eventhub  = ["10.230.6.64/26"]
 devops_service_connection_object_id = "239c15f9-6d56-4b9e-b08d-5f7779446174"
 
 db_sku_name                     = "GP_Gen5_32"
```
Comment (pasqualedevita) [inline src/env/prod/terraform.tfvars:219]:
```suggestion
db_sku_name                     = "GP_Gen5_16"
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]
@uolter we need to find a way to use only flyaway or a manual SQL script to update user passwords and dblinks.
And most important do not delete PROD-CSTAR flyaway scripts because we need they to create a bpd database from scratch

File: src/env/prod/terraform.tfvars (hunk 1 above)
Comment (uolter) [reply src/env/prod/terraform.tfvars:219]: it's in here: https://github.com/pagopa/cstar-infrastructure/pull/133

**Comment by uolter** [comment]
> @uolter we need to find a way to use only flyaway or a manual SQL script to update user passwords and dblinks.
> And most important do not delete PROD-CSTAR flyaway scripts because we need they to create a bpd database from scratch

indeed i only use flyway.  the 07-<something>.sh script it's not required. Already deleted.

File: src/psql/migrations/PROD-CSTAR/postgres/V1__update_user_password.sql [hunk 2]
```diff
@@ -0,0 +1,7 @@
+-- Update users password.
+--
+
+ALTER USER "BPD_USER" WITH PASSWORD '${bpdUserPassword}';
+ALTER USER "RTD_USER" WITH PASSWORD '${rtdUserPassword}';
+ALTER USER "FA_USER" WITH PASSWORD '${faUserPassword}';
+ALTER USER "MONITORING_USER" WITH PASSWORD '${monitoringUserPassword}';
```
Comment (pasqualedevita) [inline src/psql/migrations/PROD-CSTAR/postgres/V1__update_user_password.sql:7]:
```suggestion
ALTER USER "MONITORING_USER" WITH PASSWORD '${monitoringUserPassword}';
ALTER USER "FA_PAYMENT_INSTRUMENT_REMOTE_USER" WITH PASSWORD '${faPaymentInstrumentRemoteUserPassword}';
ALTER USER "BPD_PAYMENT_INSTRUMENT_REMOTE_USER" WITH PASSWORD '${bpdPaymentInstrumentRemoteUserPassword}';
ALTER USER "BPD_AWARD_PERIOD_REMOTE_USER" WITH PASSWORD '${bpdAwardPeriodRemoteUserPassword}';
ALTER USER "BPD_WINNING_TRANSACTION_REMOTE_USER" WITH PASSWORD '${bpdWinningTransactionRemoteUserPassword}';
ALTER USER "DASHBOARD_PAGOPA_USER" WITH PASSWORD '${dashboardPagopaUserPassword}';
ALTER USER "tkm_acquirer_manager" WITH PASSWORD '${tkmAcquirerManagerUserPassword}';
```

Comment (pasqualedevita) [reply src/psql/migrations/PROD-CSTAR/postgres/V1__update_user_password.sql:7]: add missing users

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

**Comment by pasqualedevita** [comment]
I think that we can run a flyaway baseline to set our database version to V3 without delete previous .sql scripts.
After flyaway baseline run we can run V4 sql script to update user passwords and dblinks.

https://flywaydb.org/documentation/command/baseline

https://flywaydb.org/documentation/configuration/parameters/baselineVersion

I never tested flyaway baseline but the command to run should be:

```sh
bash flyway.sh baseline "${subscription}" bpd -baselineVersion=3
```

File: src/psql/migrations/PROD-CSTAR/bpd/V1__change_db_link.sql [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+--
+
+CREATE EXTENSION IF NOT EXISTS hypopg WITH SCHEMA public;
+
+
+--
+-- Name: EXTENSION hypopg; Type: COMMENT; Schema: -; Owner: 
+--
+
+COMMENT ON EXTENSION hypopg IS 'Hypothetical indexes for PostgreSQL';
```
Comment (pasqualedevita) [inline src/psql/migrations/PROD-CSTAR/bpd/V1__change_db_link.sql:37]:
```suggestion
```
We need this?

**Review by pasqualedevita — APPROVED** [review]

**Pull request merged**